docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,298 @@
+
+```wing
+bring cloud;
+bring clients;
+
+let clientContext = new cloud.Bucket() as "ctx";
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  contextResolver: inflight () => {
+    return Json.parse(clientContext.get("ctx"));
+  },
+);
+
+// hydrates the context during deployment
+clientContext.addObject("ctx", Json.stringify(client.context))
+```
+
+In cases where reading from the filesystem is difficult and the overhead or a runtime lookup is too high, the context can be passed in via an environment variable:
+
+```wing
+bring cloud;
+bring clients;
+bring util;
+
+let bucket = new cloud.Bucket() as "shared";
+
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  contextResolver: inflight () => {
+    let ctxRaw = util.getEnv("WING_CLIENT_CONTEXT");
+    let ctxDecoded = util.fromBase64(ctxRaw);
+    return Json.parse(ctxDecoded);
+  },
+);
+
+client.add(lift { bucket: [put] });
+```
+
+Regardless of the approach used to resolve the context, using the generated client will work the same way:
+
+```ts
+import { Client } from "@acme/infra-client";
+
+// performs the context resolution
+const client = new Client();
+
+await client.bucket.put("key", "value");
+```
+
+The client code is the same as the generated javascript as if it were used within wing itself!
+
+## Implementation Details
+
+### Client context data
+
+The client context data is a JSON object that contains all the necessary information to connect to the resources.
+For now, the only information that needs to be stored are environment variables.
+This corresponds to the only API available on `IInflightHost`, which is `addEnvironment`, which is how these values are added:
+
+```json
+{
+  "version": "1",
+  "env": {
+    "BUCKET_HANDLE_a107cec5": "abcxyz"
+  }
+}
+```
+
+The client itself is effectively a proxy of an unseen/unavailable `IInflightHost`, so any expressivity that interface has must be handled by the client.
+
+### Permissions
+
+The client does not directly/automatically handle permissions needed for the arbitrary host it could run on.
+There is no expectation that the deployer of the host is also using wing so there is no shared system to transfer permission knowledge.
+
+For example, when using
+
+```wing
+client.lift(lift { bucket: [put] });
+```
+
+it would be helpful to retrieve the necessary information about the `bucket` and the `put` operation.
+
+If you're using terraform and deploying to AWS, you will need the bucket name and the correspond IAM policy.
+There is no abstract API available to store/retrieve this information currently as this is handled only by target-specific subtypes of `IInflightHost`.
+
+To keep this simple, we can provide and a helper API in the shared aws module (and other platforms, as needed) to generate the necessary policies by exposing the lift information (i.e. the `_liftMap`) from the client.
+
+```wing
+bring aws;
+
+let policies = aws.generatePolicyStatements(client.lifts);
```

Comment:
I believe this functionality basically exists already, so the only real work here is to expose it